src/net.h

```diff
@@ -64,6 +64,8 @@ static const int MAX_ADDNODE_CONNECTIONS = 8;
 static const int MAX_BLOCKS_ONLY_CONNECTIONS = 2;
 /** Maximum number of feeler connections */
 static const int MAX_FEELER_CONNECTIONS = 1;
+/** Average delay between updating address responses */
+static constexpr std::chrono::hours AVG_ADDRESS_RESPONSE_UPDATE_INTERVAL{24};
```

Comment:
I would say 24h is pretty arbitrary.

Upper bound:
Shouldn't be too long so that the returned records are still relevant for GETADDR requestor. Although currently `IsTerrible` is fine with everything in last 30 days, we also should make sure that nodes in the response are not dead. I can't find the data, but I remember reading that the churn among reachable nodes (the ones we care) is low, so 1 day should be fine.

Lower bound:
Too frequent cache updates can leak privacy again. For example, think about the attack where you map onion to ipv4 of the same node by correlating addrmans. Ideally, the response in two caches does not intersect.

The main event for both bounds is probably a feeler (because it updated nTime of the feeler's addr record). Feeler every 2 minutes update 720 records a day, so that's the same order as GETADDR response (1000 records).

Same order means that not many of the new records will be missed in the cached response , as well as updates (upper bound concern).  It also means that the caches are updated at the same pace as the AddrMan updates happen (lower bound concern).

I think any reasoning beyond that requires substantial research into attacks.

Reply:
We should dissociate network _churn_, i.e how many peers fall offline after X and peers _rotation_, i.e changes among our connections demography after same X. I think for counter-measure to be efficient you need rotation to be high enough to blur adversary guess of your peers topology. 

Obviously, a higher rotation target is likely to come with a higher churn, and therefore decrease validity of returned addresses. But I think it's okay to try more connections compare to leaking your topology, goals don't have same value. 

I agree, sounds hard to guess the sweet spot with further research, you should comment the trade-off and invite to further investigation.

With regards to bounds, I think it make sense to evaluate that after a feeler cycle of a day, you may dispose from a more accurate view of the network and therefore you should update your cache. An average delay of 36h would sounds correct to me too, even making counter-measure a bit more efficient.

Reply:
As for rotation of outbound peers, I don't know if we actually have any mechanism to do that programmatically. This idea was discussed [here](https://github.com/bitcoin/bitcoin/pull/4723/files) but the PR was closed as further research is needed. As rotation of inbound peers, does anyone know whether that even exists & how it is done?

Reply:
Inbound rotation is something I've never heard of :)

It practice it just means dropping an existing connection someone made to us... It doesn't feel useful from the first glance, unless that peer was malicious or broken. 
An inbound connection can't be immediately replaced, so this seems like a loose-loose w.r.t security. We can drop a spy, true, but right now that is solved by slower relays to inbounds etc.

We "rotate" inbound peers if we're at capacity (117 inbounds or whatnot), but that's a completely different story. That's called eviction.